bftengine/src/preprocessor/PreProcessor.cpp

```diff
@@ -424,18 +409,30 @@ PreProcessor::PreProcessor(shared_ptr<MsgsCommunicator> &msgsCommunicator,
   addTimers();
 }
 
-PreProcessor::~PreProcessor() {
-  LOG_TRACE(logger(), "~PreProcessor start");
-
-  stop();
+void PreProcessor::stop() {
+  LOG_DEBUG(logger(), "Going to stop PreProcessor");
+  if (!msgLoopDone_) {
```

Comment:
=== same comment to all below cases when you trigger notify_one/all on this cond variable. ===

Are you catching a lock somewhere?  If not, I think that this will never be 100% thread safe. It will be ok in most cases but fail in others (e.g once in 100 or once in 1000) , and lead to undefined behaviors in shutdown (crashes, etc).
You are changing a shared state and notifying without taking a lock. This kind of usage isn't according to the way required:
```
The thread that intends to modify the shared variable must:

1) Acquire a std::mutex (typically via [std::lock_guard](https://en.cppreference.com/w/cpp/thread/lock_guard))
2) Modify the shared variable while the lock is owned
3) Call [notify_one](https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one) or [notify_all](https://en.cppreference.com/w/cpp/thread/condition_variable/notify_all) on the std::condition_variable (can be done after releasing the lock)
```

more info in:
https://en.cppreference.com/w/cpp/thread/condition_variable

Reply:
I still think that the mutex lock is unneeded in case you work with a lock-free underlined queue. You actually have no object to guard. In terms of the guide, there is no "shared variable". The conditional variable is used in my case just to avoid constant polling of the queue in case there is no data there. notify_one could be called without locking.

Reply:
`Even if the shared variable is atomic, it must be modified while owning the mutex to correctly publish the modification to the waiting thread.`

Reply:
The condition variable is not waiting for any variable to be modified, please look again at the code.